source/parse.h

```diff
@@ -3071,16 +3071,31 @@ class parser
         -> void
     {
         auto m = std::string{msg};
-        if (include_curr_token) {
-            m += std::string(" (at '") + curr().to_string(true) + "')";
-        }
-        if (
-            err_pos == source_position{}
-            && peek(0)
-            )
-        {
-            err_pos = peek(0)->position();
+
+        if (done()) {
+            int i = 0;
+            while(!peek(i) && pos + i > 0) { --i; };
```

Comment:
I see this is making `error` more robust if we're `done()`, but I'm not sure what this loop is doing... is it assuming we might be more than one past the end? But that shouldn't happen -- see the `assert (pos <= std::ssize(*tokens));` in `done()`.

Instead of this diff, what if we changed this original code

```
        if (include_curr_token) {
            m += std::string(" (at '") + curr().to_string(true) + "')";
        }
        if (
            err_pos == source_position{}
            && peek(0)
            )
        {
            err_pos = peek(0)->position();
```

to this?

```
        auto i = done() ? -1 : 0;
        assert (pos+i < std::ssize(*tokens));
        if (include_curr_token) {
            m += std::string(" (at '") + peek(i)->to_string(true) + "')";
        }
        if (
            err_pos == source_position{}
            && peek(i)
            )
        {
            err_pos = peek(i)->position();
```

Does that accomplish what you were aiming for here?

Reply:
Your proposed code was the first I made. Later, I wondered if I could assume that going back by one would be enough. That is why I have introduced the loop to ensure to go back to the last good one. If you think that going back by one is enough, we shall add an assertion that will emphasize this assumption.

Reply:
I have the following assertion in mind:
```cpp
auto i = done() ? -1 : 0;
assert (peek(i)); // <---- this one
assert (pos+i < std::ssize(*tokens));
```

Reply:
Yes, that looks good.

Reply:
OK, I am correcting it.

Reply:
In fact asserting `peek(i)` is better, and you can remove the second assertion which is redundant (the peek checks it already, plus it checks `>0` so it's strictly better).

Reply:
As peek implementation is the following:
```cpp
    auto peek(int num) const
        -> token const*
    {
        assert (tokens);
        if (
            pos + num >= 0
            && pos + num < std::ssize(*tokens)
            )
        {
            return &(*tokens)[pos + num];
        }
        return {};
    }
```

It makes `assert (pos+i < std::ssize(*tokens));` redundant as it is already checked in `peek()`.

Reply:
Pushed corrected code.